### PR TITLE
Tweak validation code

### PR DESCRIPTION
**Pull request by vidartf**
Three main points:
- Refactor out SciType to be self contained (less of an abstract base class). Less repetition of code.
- Will not coerce `None` and `Undefined`.
- Gives a warning when coercing one ndarray to another (a copy is created).

### CODE REVIEW

**Comment by SylvainCorlay** [comment]
I am +1 on droppin python 3.3 and 3.4!
Maybe we should add 3.6 to the list though.

File: traittypes/traittypes.py [hunk 1]
```diff
@@ -22,6 +24,10 @@ class SciType(TraitType):
 
     """A base trait type for numpy arrays, pandas dataframes and series."""
 
+    def __init__(self, **kwargs):
+        super(SciType, self).__init__(**kwargs)
+        self.validators = []
```
Comment (SylvainCorlay) [inline traittypes/traittypes.py]: +1

File: traittypes/traittypes.py [hunk 2]
```diff
@@ -70,13 +85,20 @@ class Array(SciType):
     def validate(self, obj, value):
         if value is None and not self.allow_none:
             self.error(obj, value)
+        if value is None or value is Undefined:
+            return super(Array, self).validate(obj, value)
```
Comment (SylvainCorlay) [inline traittypes/traittypes.py]: +1 on this factorization.

File: traittypes/traittypes.py [hunk 3]
```diff
@@ -70,13 +85,20 @@ class Array(SciType):
...
-            for validator in self.validators:
-                value = validator(self, value)
-            return value
+            r = np.asarray(value, dtype=self.dtype)
+            if isinstance(value, np.ndarray) and r is not value:
+                warnings.warn(
+                    'Given trait value dtype "%s" does not match required type "%s". '
+                    'A coerced copy has been created.' % (
+                        np.dtype(value.dtype).name,
+                        np.dtype(self.dtype).name))
```
Comment (SylvainCorlay) [inline traittypes/traittypes.py]: Do you want this warning to occur if you do `foo.bar = [1, 2]`?

Comment (vidartf) [reply traittypes/traittypes.py]: I didn't think so (hence the test if original value is an ndarray). Do you think it should?

Comment (SylvainCorlay) [reply traittypes/traittypes.py]: Ah gotcha.

Comment (SylvainCorlay) [reply traittypes/traittypes.py]: No, I agree with you.

**Comment by SylvainCorlay** [comment]
This is all-good to me. Do you want to remove the failing test from the matrix?

**Comment by vidartf** [comment]
Pushed and passing!

**Comment by SylvainCorlay** [comment]
Thanks!

Do you think that we need more before tagging a new release?

**Pull request merged**